File: Bit Manipulation/898. Bitwise ORs of Subarrays/solution.py
class Solution:
    def subarrayBitwiseORs(self, arr: List[int]) -> int:
        n = len(arr)
        res = set()
        dp = set()
        for i in range(n):
            tmp = dp.copy()
            
            dp.clear()
            dp.add(arr[i])
            for prev in tmp:
                dp.add(prev | arr[i])
            
            res.update(dp)
        return len(res)

File: Bit Manipulation/898. Bitwise ORs of Subarrays/solution.cpp
class Solution {
public:
    int subarrayBitwiseORs(vector<int>& arr) {
        unordered_set<int> ans;
        unordered_set<int> curr, next;
        
        for (int& num : arr) {
            next = {num};
            for (int prev : curr) {
                next.insert(prev | num);
            }
            curr = next;

            ans.insert(curr.begin(), curr.end());
        }

        return ans.size();
    }
};